lib/DDGC/Schema/Result/User/Subscription.pm

```diff
@@ -0,0 +1,89 @@
+package DDGC::Schema::Result::User::Subscription;
+# ABSTRACT: Activity feed subscription result class
+
+use Moo;
+extends 'DDGC::Schema::Result';
+use DBIx::Class::Candy;
+
+table 'user_subscription';
+
+column id => {
+    data_type => 'bigint',
+    is_auto_increment => 1,
+};
+primary_key 'id';
+
+column users_id => {
+    data_type   => 'bigint',
+    is_nullable => 0,
+};
+
+column category => {
+    data_type => 'text',
+    is_nullable => 0,
+};
+
+column action => {
+    data_type => 'text',
+    is_nullable => 1,
+};
+
+column meta1 => {
+    data_type => 'text',
+    is_nullable => 1,
+};
+
+column meta2 => {
+    data_type => 'text',
+    is_nullable => 1,
+};
+
+column meta3 => {
+    data_type => 'text',
+    is_nullable => 1,
+};
+
+belongs_to 'user', 'DDGC::Schema::Result::User', 'users_id';
+
+has_many 'activity', 'DDGC::Schema::Result::ActivityFeed', sub {
+    my ( $args ) = @_;
+
+    my $activity = $args->{foreign_alias};
+    my $subscription = $args->{self_alias};
+
+    +{
+        -and => {
+            "$activity.category" => "$subscription.category",
+            -or => [
+                "$subscription.action" => { '=' => undef },
+                "$activity.action" => "$subscription.action",
+            ],
+            -or => [
+                "$subscription.meta1" => { '=' => undef },
+                "$activity.meta1" => { 'like' => "$subscription.meta1" },
+            ],
+            -or => [
+                "$subscription.meta2" => { '=' => undef },
+                "$activity.meta2" => { 'like' => "$subscription.meta2" },
+            ],
+            -or => [
+                "$subscription.meta3" => { '=' => undef },
+                "$activity.meta3" => { 'like' => "$subscription.meta3" },
+            ],
+            -or => [
+                "$activity.for_user" => { '=' => undef },
+                "$activity.for_user" => { '=' => "$subscription.users_id" },
+            ],
+            -or => [
+                "$activity.for_role" => { '=' => undef },
+                "$activity.for_role" => {
+                    -in => \"SELECT role FROM user_role WHERE users_id = $subscription.users_id",
```

Comment:
So this really, really looks like something that should be a bound parameter prepared statement, right? A glance at this here interpolation / literal-expansion-in-SQL-stringery should rightly make you scream.

...except `$subscription` is a table alias and not some arbitrary value. I'm pretty sure this is, if not The Right Thing, fairly safe.